src/etos_lib/kubernetes/etos.py

```diff
@@ -0,0 +1,191 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Kubernetes client for ETOS custom resources."""
+import os
+import logging
+from pathlib import Path
+from contextlib import contextmanager
+from typing import Optional
+from pydantic import BaseModel
+from kubernetes import config
+from kubernetes.client import api_client
+from kubernetes.client.exceptions import ApiException
+from kubernetes.dynamic import DynamicClient
+from kubernetes.dynamic.resource import Resource as DynamicResource, ResourceInstance
+from kubernetes.dynamic.exceptions import (
+    ResourceNotFoundError,
+    ResourceNotUniqueError,
+    NotFoundError,
+)
+
+config.load_config()
+NAMESPACE_FILE = Path("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+
+
+class NoNamespace(Exception):
+    """NoNamespace exception is raised when ETOS could not determine the current namespace."""
+
+
+class Resource:
+    """Resource is the base resource client for ETOS custom resources."""
+
+    client: DynamicResource
+    namespace: str = "default"
+    strict: bool = False
+    __cache = {}
+
+    @contextmanager
+    def _catch_errors_if_not_strict(self):
+        """Catch errors if the strict flag is False, else raise."""
+        try:
+            yield
+        # Internal exceptions
+        except NoNamespace:
+            if self.strict:
+                raise
+        # Client exceptions
+        except ApiException:
+            if self.strict:
+                raise
+        # Dynamic client exceptions
+        except (ResourceNotFoundError, ResourceNotUniqueError):
+            if self.strict:
+                raise
+        # Built-in exceptions
+        except AttributeError:
+            # AttributeError happens if ResourceNotFoundError was raised when setting up
+            # clients.
+            if self.strict:
+                raise
+
+    def __full_resource_name(self, name: str):
+        """Full resource name will return the group, version, namespace and kind."""
+        return (
+            f"{self.client.group}/{self.client.api_version}/{self.client.kind} "
+            f"{self.namespace}/{name}"
+        )
+
+    def get(self, name: str, cache=True) -> Optional[ResourceInstance]:
+        """Get a resource from Kubernetes by name.
+
+        if Cache is set to False, then make sure to get the resource from kubernetes,
+        if Cache is True, then the cache will be used every time.
+        There is no cache invalidation!
```

Comment:
Risk for outdated data? Maybe set a time interval how often the cache is allowed to be.

Reply:
Yes. Another risk is memory usage. I will just remove the caching entirely.